Add SwiftUI enhancements

jsflax opened the pull request:
This PR adds the `@StateRealmObject`, `@ObservedObject`, and `@ObservedResults` property wrappers, which will allow users to more easily wield thread confined live types: `Objects`, `EmbeddedObjects`, `Lists`, or `Results`:

```swift
struct PersonView: View {
    @ObservedResults(Person.self) var results

    var body: some View {
        return NavigationView {
            List {
                ForEach(results) { person in
                    NavigationLink(destination: PersonDetailView(person: person)) {
                        Text(person.name)
                    }
                }
                .onDelete(perform: $results.remove)
            }
            .navigationBarTitle("People", displayMode: .large)
            .navigationBarItems(trailing: Button("Add") {
                $results.append(Person())
            })
        }
    }
}
```

The wrapper allows properties to be wielded normally when doing most operations, and adds bonus convenience methods for under-the-hood write transactions for each type (and their properties).

jsflax commented on RealmSwift/SwiftUI.swift:
```diff
@@ -0,0 +1,349 @@
+                guard let (realm, resolved) = resolve(wrappedValue) else {
+                    return
+                }
+                try! realm.write {
```
The setter cannot throw because it is used to set the wrappedValue with the nonmutating setter. What's the alternative?

jsflax replied:
Forgot to, but I still need to do this even though this comment is outdated.

leemaguire commented on SwiftUITests/SwiftUITests.swift:
```diff
@@ -0,0 +1,114 @@
+//
```
needs a licence header

leemaguire commented on examples/ios/swift/ListSwiftUI/Views/ContentView.swift:
```diff
@@ -1,126 +1,148 @@
+        List {
+            // Using the `$` will bind the Dog List to the view
+            ForEach(dogs) { dog in
+                TextField("dog name", text: bind(dog, \.name))
```
will this `bind` API be hard to discover? Do we also want to extend `View` to do the also?

jsflax replied:
My other thought was to put `bind` on `ThreadConfined` types, but I'm not sure if there is value in that outside of SwiftUI (it's API bloat otherwise, but maybe that's fine?). I'm open to ideas anyways.

leemaguire replied:
I'm just thinking we should extend our own domain and not another party's. But it's tricky because its SwiftUI specific. I guess I'm fine with extending `View` as its a very niche case

jsflax replied:
I actually don't think this is terribly niche, which is why I'm trying to be careful– binding to a SwiftUI view will likely be fairly common (TextFields, sliders, etc.).

Thoughts @tgoyne ?

jsflax commented:
@tgoyne I've diverged from Eric's branch and will changeover to master once Eric's changes are merged, but could use another set of eyes on the actual swiftUI related changes.

tgoyne commented on Configuration/SwiftUITestHost.xcconfig:
```diff
@@ -0,0 +1,15 @@
+OTHER_LDFLAGS[sdk=macosx*] = -framework SwiftUI;
+LD_RUNPATH_SEARCH_PATHS[sdk=macosx*] = $(inherited) @executable_path/../Frameworks @loader_path/../Frameworks;
+
+MACOS_DEPLOYMENT_TARGET = 11.0;
```
This needs to target 10.15 to be able to run on CI.

tgoyne commented on Realm.xcodeproj/project.pbxproj:
```diff
@@ -2850,9 +2853,15 @@
 				DEBUG_INFORMATION_FORMAT = dwarf;
+				DEVELOPMENT_ASSET_PATHS = "";
+				DEVELOPMENT_TEAM = "";
+				ENABLE_HARDENED_RUNTIME = YES;
```
This is why you have to do awkward things with code signing and it should just be off unless it's actually required.

tgoyne commented on RealmSwift/List.swift:
```diff
@@ -615,23 +615,33 @@ extension List: MutableCollection {
-            if o >= destination {
-                o += tmp.count
+        if Element.self is EmbeddedObject.Type {
+            for offset in offsets {
```
Is there some reason not to use this code path for non-embedded objects as well?

jsflax replied:
It seemed like the way for non-embedded objects was "more correct," and that this was a workaround, but I can check if that actually matters

jsflax replied:
I've added a test for this and collapsed the conditional.

tgoyne commented on RealmSwift/SwiftUI.swift:
```diff
@@ -21,13 +21,31 @@ import SwiftUI
+            return lastValue
+        }
+        if parent.isFrozen {
+            parent = try! Realm(configuration: parent.realm!.configuration).thaw(parent)!
```
Can't this just be `parent.realm!.thaw(parent)`?

The force-unwrap is unsafe and this'll crash if the object was deleted after the version which the parent was frozen at.

tgoyne commented on RealmSwift/SwiftUI.swift:
```diff
@@ -142,130 +131,103 @@ extension RealmBinding: RandomAccessCollection, BidirectionalCollection, Collect
         get {
-            self.wrappedValue[position]
+            if self.get().isFrozen { throwRealmException("Cannot get values from a frozen realm") }
+            return self.get()[position]
```
get() can do some nontrivial work so this should avoid calling it twice in one function.

tgoyne commented on RealmSwift/SwiftUI.swift:
```diff
@@ -300,50 +262,96 @@ private final class Box<T: RealmSubscribable & ThreadConfined>: ObservableObject
-            let results = filter == nil ? realm.objects(U.self) : realm.objects(U.self).filter(filter!)
-            _box = StateObject(wrappedValue: Box(results.freeze()))
-        }
+        let actualRealm = realm == nil ? try! Realm(configuration: Realm.Configuration.defaultConfiguration) : realm!
```
```suggestion
        let actualRealm = realm ?? try! Realm()
```

iandundas commented on RealmSwift/SwiftUI.swift:
```diff
@@ -0,0 +1,700 @@
+    public var wrappedValue: ObjectType {
+        get {
+            if storage.value.realm == nil {
+                // if unmangaged, return the unmanaged value
```
small typo:
```suggestion
                // if unmanaged, return the unmanaged value
```

jsflax commented:
Some changes since last review:
- KVO has been added for unmanaged objects
- Optionals are now supported for `@StateRealmObject`
- `@StateRealmObject` is now exclusively for `List`, `Objects`/`EmbeddedObjects`, and `Optional` variants of the previously mentioned types. It is meant to be used in the same way as SwiftUI's `@StateObject`.
- `@ObservedRealmObject` is now exclusively for `List`, `Objects`/`EmbeddedObjects`. It is meant to be used in the same way as SwiftUI's `@ObservedObject`.
- `@FetchRealmResults` has been added to better support binding Results types to a view while allowing dynamic querying and multiple Realm support. The default realm configuration for a `@FetchRealmResults` is passed in through the environment of a view hierarchy, e.g.,:
```swift
struct View2: View {
    // the realm used here will be the realm opened from
    // `myCustomConfiguration`
    @FetchRealmResults(MyObject.self) var results

    ...
}

struct View1: View {
    var body: some View {
        View2().environment(\.realmConfiguration, myCustomConfiguration)
    }
}
```
- The example in the examples directory has been changed to the Reminders application

TODO:
- Need to update the README.
- Add one example showing multiple realm usage

@tgoyne can you PTAL?

jsflax commented:
@tgoyne I've resolved all previous comments, just wrestling with the CI now.

leemaguire commented on CHANGELOG.md:
```diff
@@ -1,6 +1,19 @@
+* Add support for "thawing" objects. `Realm`, `Results`, `List` and `Object` now
+have `thaw()` methods which return a live copy of the frozen object. This enables app behvaior where a frozen object can be made live again in order to mutate values. For example, first freezing an object passed into UI view, then thawing the object in the view to update values.
+* Add `@StateRealmObject` for SwiftUI support. This property wrapper type instantiates an observable object on a View. Use in place of `SwiftUI.StateObject` for Realm `Object`, `List`, and `EmbeddedObject` types.
+* Add `@ObservedObject` for SwiftUI support. This property wrapper type subscribes to an observable object and invalidates a view whenever the observable object changes. Use in place of `SwiftUI.ObservedObject` for Realm `Object`, `List`, or `EmbeddedObject` types.
```
should this be `@ObservedRealmObject`?

jsflax replied:
Yes

tgoyne commented on CHANGELOG.md:
```diff
@@ -1,11 +1,18 @@
+* Add `Object.bind` and `EmbeddedObject.bind` for SwiftUI support. This allows you to create bindings of realm properties when a propertyWrapper is not available for you to do so, e.g., `TextField("name", personObject.bind(\.name))`.
+
+### Fixed
+* <How to hit and notice issue? what was the impact?> ([#????](https://github.com/realm/realm-cocoa/issues/????), since v?.?.?)
```
Incorrect merge.

jsflax replied:
fixed

tgoyne commented on RealmSwift/SwiftUI.swift:
```diff
@@ -0,0 +1,619 @@
+        var sortDescriptor: SortDescriptor? {
+            willSet {
+                if let sortDescriptor = newValue {
+                    value = baseValue.sorted(byKeyPath: sortDescriptor.keyPath, ascending: sortDescriptor.ascending)
```
Rather than duplicating all this logic in a whole bunch of places there should just be a single function that updates things which is called in each didSet.

tgoyne commented on RealmSwift/SwiftUI.swift:
```diff
@@ -0,0 +1,619 @@
+            value.realm?.beginWrite()
+        }
+        block()
+        try? value.realm?.commitWrite()
```
try? is bad and this should only commit if it started the write

tgoyne commented on RealmSwift/Tests/ListTests.swift:
```diff
@@ -301,6 +301,42 @@ class ListTests: TestCase {
         assertThrows(array.move(from: 2, to: 0))
     }
 
+    func testMoveFromOffsets() {
```
This should also test zero offsets and more than one.

tgoyne commented on examples/installation/osx/swift/DynamicExample/DynamicExample.xcodeproj/project.pbxproj:
```diff
@@ -305,7 +305,7 @@
 				ENABLE_STRICT_OBJC_MSGSEND = YES;
 				ENABLE_TESTABILITY = YES;
-				FRAMEWORK_SEARCH_PATHS = "../../../realm-swift-latest/osx/swift-11.3";
+				FRAMEWORK_SEARCH_PATHS = "../../../realm-swift-latest/12.3/osx/swift-11.3";
```
This path doesn't exist.

tgoyne commented on scripts/pr-ci-matrix.rb:
```diff
@@ -47,6 +47,7 @@ def minimum_version(major)
   'cocoapods-ios-dynamic' => oldest_and_latest,
   'cocoapods-watchos' => oldest_and_latest,
   # 'cocoapods-catalyst' => oldest_and_latest,
+  'swiftui' => latest,
```
You also need to run this script to populate the file that jenkins actually looks at.

jsflax commented on RealmSwift/EmbeddedObject.swift:
```diff
@@ -165,62 +165,6 @@ extension EmbeddedObject: RealmCollectionValue {
         return RLMDynamicGet(self, prop)
     }
 
-    // MARK: Notifications
```
There doesn't seem to be a clean way to do this from an extension.

tgoyne replied:
The public thing needs to be on Object/EmbeddedObject which then could call an internal function on the ObjectBase extension rather than putting the public thing on ObjectBase.

jsflax commented on Realm/Tests/SwiftUITestHostUITests/SwiftUITestHostUITests.swift:
```diff
@@ -0,0 +1,168 @@
+    var realm: Realm!
+    let app = XCUIApplication()
+
+    override func setUpWithError() throws {
```
I excluded the file using swift version guards and made the tests themselves guarded by the `available` directive.

jsflax commented:
@tgoyne , addressed all comments– still haven't come up with a better name than `FetchRealmResults` though. Added a conditional in `Realm.add` that checks if SwiftUI KVO is active on an unmanaged objects and removes if so. Added a test for this as well. PTAL.

tgoyne commented on RealmSwift/SwiftUI.swift:
```diff
@@ -121,17 +133,12 @@ private final class ObservableStoragePublisher<ObjectType>: Publisher where Obje
-                value.addObserver(kvo, forKeyPath: name, options: .new, context: nil)
+            for property in schema.properties {
+                keyPaths.append(property.name)
+                value.addObserver(kvo, forKeyPath: property.name, options: .new, context: nil)
```
This shouldn't be asking for the new value as it doesn't use it.

tgoyne commented on RealmSwift/SwiftUI.swift:
```diff
@@ -259,56 +266,37 @@ private class ObservableStorage<ObservedType>: ObservableObject where ObservedTy
+            if let sortDescriptor = sortDescriptor {
+                value = value.sorted(byKeyPath: sortDescriptor.keyPath, ascending: sortDescriptor.ascending)
+            } else {
+                value = value.sorted(by: [])
```
Why?

tgoyne commented on scripts/.jenkins.yml:
```diff
@@ -0,0 +1,281 @@
+
+# Yaml Axis Plugin
```
This is the wrong path for this file.

tgoyne reviewed (commented):
changelog merge is still incorrect

jsflax commented:
I fixed the CHANGELOG (🤞), renamed `FetchRealmResults` to `ObservedResults`, renamed `KVO` to `SwiftUIKVO`, moved `.jenkins.yml` file to root, and added an `_objectSchema` method to `ObjectBase` as that had been missing.

tgoyne commented:
There isn't really a reason to add _objectSchema; just using RLMObjectBaseObjectSchema() directly is basically the same code but with less copying involved.

jsflax commented:
I'll revert that. I'm having a bit of difficulty understanding why the swiftly tests continue to fail– https://ci.realm.io/job/objc_pr/target=swiftpm,xcode_version=12.1/7669/console

Any thoughts on why the guards I've added aren't sufficient?

tgoyne commented:
StateObject requires Xcode 12.2, which is Swift 5.3.1

jsflax commented:
<img width="242" alt="Screenshot 2021-02-12 at 18 38 56" src="https://user-images.githubusercontent.com/4356928/107833531-91c82380-6d61-11eb-8c95-5698ef7a6a0a.png">

@tgoyne our first passing UI tests 😄 . Let me know if you have any more thoughts

tgoyne commented on build.sh:
```diff
@@ -668,7 +668,9 @@ case "$COMMAND" in
 
     "test-swiftui")
-        xctest "-scheme SwiftUITestHost -configuration $CONFIGURATION -sdk iphonesimulator -destination 'name=iPhone 12'"
+        sh build.sh download-core
```
This should be in the scheme's pre-build action, not here.

tgoyne commented on Configuration/SwiftUITestHost.xctestplan:
```diff
@@ -0,0 +1,24 @@
+    }
+  ],
+  "defaultOptions" : {
+    "mainThreadCheckerEnabled" : false
```
We want the main thread checker enabled for this (and UI tests are an example of a case where it might even be useful). It's only disabled in the Xcode 11 version of the test plans, not the default one.

tgoyne commented on Realm/Tests/SwiftUITestHost/Objects.swift:
```diff
@@ -0,0 +1,48 @@
+@objcMembers class ReminderList: Object, ObjectKeyIdentifiable {
+    dynamic var name = "New List"
+    dynamic var icon: String = "list.bullet"
+    var reminders = RealmSwift.List<Reminder>()
```
```suggestion
    let reminders = RealmSwift.List<Reminder>()
```

jsflax replied:
I've had to unapply this suggestion– to be able to use the `$` convenience methods, it needs to be mutable.

tgoyne commented on Realm/Tests/SwiftUITestHost/SwiftUITestHostApp.swift:
```diff
@@ -0,0 +1,255 @@
+            }).accessibilityIdentifier("picker")
+        }
+        .navigationTitle(reminder.title)
+        .navigationBarItems(trailing:
+        Button("Save") {
```
```suggestion
        .navigationBarItems(trailing: Button("Save") {
```

tgoyne commented on Realm/Tests/SwiftUITestHost/SwiftUITestHostApp.swift:
```diff
@@ -0,0 +1,255 @@
+    @State var showReminderForm = false
+
+    func shouldFocusReminder(_ reminder: Reminder) -> Bool {
+        return newReminderAdded &&
+            list.reminders.lastIndex(of: reminder) == (list.reminders.count - 1)
```
```suggestion
        return newReminderAdded && list.reminders.last == reminder
```

tgoyne commented on RealmSwift/Object.swift:
```diff
@@ -212,11 +212,10 @@ extension Object: RealmCollectionValue {
+     Notifications are delivered via the standard run loop, and so can't be
+     delivered while the run loop is blocked by other activity. When
+     notifications can't be delivered instantly, multiple notifications may be
+     coalesced into a single notification.
```
The previous version of this wasn't great, but removing the mention of the queue isn't an improvement to it.

tgoyne commented on RealmSwift/SwiftUI.swift:
```diff
@@ -0,0 +1,599 @@
+private func createBinding<T: ThreadConfined, V>(_ value: T,
+                                                 forKeyPath keyPath: ReferenceWritableKeyPath<T, V>) -> Binding<V> {
+    if value.isFrozen {
+        throwRealmException("Should not bind frozen value")
```
This appears to be something that's supposed to be unreachable. If that's correct, it should be a `precondition()` to make that clear, while if it's something the user actually can hit then it should be a bit more descriptive.

tgoyne commented on RealmSwift/SwiftUI.swift:
```diff
@@ -0,0 +1,599 @@
+    }
+    // store last known value outside of the binding so that we can reference it if the parent
+    // is invalidated
+    let lastValue = value[keyPath: keyPath]
```
Shouldn't lastValue be updated each time the value changes? If that doesn't make sense to do then it should be `initialValue`.

tgoyne commented on RealmSwift/SwiftUI.swift:
```diff
@@ -0,0 +1,599 @@
+    public mutating func update() {
+        // When the view updates, it will inject the @Environment
+        // into the propertyWrapper
+        if storage.configuration == nil ||
```
Making `Realm.Configuration` conform to Equatable and putting this logic there seems better.

tgoyne commented on RealmSwift/SwiftUI.swift:
```diff
@@ -0,0 +1,599 @@
+
+    /// :nodoc:
+    func remove<V>(at index: Index) where Value == List<V> {
+        let list = self.wrappedValue.realm != nil ?
```
This logic shows up in a lot of places and would be worth extracting to a function.

tgoyne commented on RealmSwift/Tests/SwiftUITests.swift:
```diff
@@ -0,0 +1,239 @@
+    // MARK: - ObservedResults Operations
+    func testResultsAppendUnmanagedObject() throws {
+        let object = SwiftUIObject()
+        let fullResults = ObservedResults(SwiftUIObject.self,
```
Indentation is off in a bunch of spots after the renames.

tgoyne reviewed (commented):
Took another pass over the whole thing. Mostly looks good and all my remaining comments are minor things.
